pkg/storage/ingest/reader.go

```diff
@@ -73,7 +73,10 @@ type PartitionReader struct {
 	partitionID   int32
 	consumerGroup string
 
-	client *kgo.Client
+	client  *kgo.Client
+	fetcher interface {
+		pollFetches(context.Context) kgo.Fetches
+	}
```

Comment:
[nit] is this a pattern we commonly use? Would rather see the interface as a separate type.

Reply:
After reading more of the code I found this is a bit simpler to understand:

```
diff --git a/pkg/storage/ingest/reader.go b/pkg/storage/ingest/reader.go
index d8f826118..0c63ac245 100644
--- a/pkg/storage/ingest/reader.go
+++ b/pkg/storage/ingest/reader.go
@@ -65,6 +65,8 @@ type recordConsumer interface {
 	consume(context.Context, []record) error
 }
 
+type pollingFetcherFunc func(context.Context) kgo.Fetches
+
 type PartitionReader struct {
 	services.Service
 	dependencies *services.Manager
@@ -73,10 +75,8 @@ type PartitionReader struct {
 	partitionID   int32
 	consumerGroup string
 
-	client  *kgo.Client
-	fetcher interface {
-		pollFetches(context.Context) kgo.Fetches
-	}
+	client      *kgo.Client
+	pollingFunc pollingFetcherFunc
 
 	newConsumer consumerFactory
 	metrics     readerMetrics
@@ -143,7 +143,6 @@ func newPartitionReader(kafkaCfg KafkaConfig, partitionID int32, instanceID stri
 		logger:                log.With(logger, "partition", partitionID),
 		reg:                   reg,
 	}
-
 	r.Service = services.NewBasicService(r.start, r.run, r.stop)
 	return r, nil
 }
@@ -187,12 +186,19 @@ func (r *PartitionReader) start(ctx context.Context) (returnErr error) {
 	}
 
 	if r.kafkaCfg.ReplayConcurrency > 1 {
-		r.fetcher, err = newConcurrentFetchers(ctx, r.client, r.logger, r.kafkaCfg.Topic, r.partitionID, startOffset, r.kafkaCfg.ReplayConcurrency, r.kafkaCfg.RecordsPerFetch, &r.metrics)
+		fetcher, err := newConcurrentFetchers(ctx, r.client, r.logger, r.kafkaCfg.Topic, r.partitionID, startOffset, r.kafkaCfg.ReplayConcurrency, r.kafkaCfg.RecordsPerFetch, &r.metrics)
 		if err != nil {
 			return errors.Wrap(err, "creating concurrent fetchers")
 		}
+
+		r.pollingFunc = fetcher.pollFetches
 	} else {
-		r.fetcher = r
+		r.pollingFunc = func(ctx context.Context) kgo.Fetches {
+			defer func(start time.Time) {
+				r.metrics.fetchWaitDuration.Observe(time.Since(start).Seconds())
+			}(time.Now())
+			return r.client.PollFetches(ctx)
+		}
 	}
 
 	// Enforce the max consumer lag (if enabled).
@@ -242,7 +248,7 @@ func (r *PartitionReader) run(ctx context.Context) error {
 }
 
 func (r *PartitionReader) processNextFetches(ctx context.Context, delayObserver prometheus.Observer) error {
-	fetches := r.fetcher.pollFetches(ctx)
+	fetches := r.pollingFunc(ctx)
 	r.recordFetchesMetrics(fetches, delayObserver)
 	r.logFetchErrors(fetches)
 	fetches = filterOutErrFetches(fetches)
@@ -716,13 +722,6 @@ func (r *PartitionReader) waitReadConsistency(ctx context.Context, withOffset bo
 	return err
 }
 
-func (r *PartitionReader) pollFetches(ctx context.Context) (result kgo.Fetches) {
-	defer func(start time.Time) {
-		r.metrics.fetchWaitDuration.Observe(time.Since(start).Seconds())
-	}(time.Now())
-	return r.client.PollFetches(ctx)
-}
-
 // fetchWant represents a range of offsets to fetch.
 // Based on a given number of records, it tries to estimate how many bytes we need to fetch, given there's no support for fetching offsets directly.
 // fetchWant also contains the channel on which to send the fetched records for the offset range.

```

WDYT?

Reply:
I removed the anonymous interface in 9f27ea065a18a6122e84c0cd5f881b1c93b894ed. Looks good?